kame/kame/dsadd/dsadd.8
.\"	$KAME: dsadd.8,v 1.2 2003/01/10 06:12:45 sakane Exp $
.\"
.\" Copyright (C) 2002 WIDE Project.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the project nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd December 27, 2002
.Dt DSADD 8
.Os KAME
.\"
.Sh NAME
.Nm dsadd
.Nd Dead Security Association Detection Daemon
.\"
.Sh SYNOPSIS
.Nm dsadd
.Bk -words
.Op Fl c Ar number
.Ek
.Bk -words
.Op Fl i Ar time
.Ek
.Bk -words
.Op Fl LESvd
.Ek
.\"
.Sh DESCRIPTION
.Nm
detects and deletes the dead security association.
.Nm
searches the security association database in the kernel periodically,
then sends ICMP echo requests to the peer of each SA.
if any ICMP echo reply won't be back,
.Nm
will delete the SA immediately.
.Bl -tag -width Ds
.It Fl c
specify the number of icmp echo packet to be sent.
default is 2 times.
.It Fl L
force to check only once.
.It Fl i
(a) if the current round takes more than the specified
time in seconds the next round will start immediately
after the current will finish.  (b) otherwise the next
will start after the specified time from the current
first check.
default is 300 seconds.
.Bd -unfilled -literal -offset indent
(a)
+----- first round -----+---- second round --...
+-- specified time --+

(b)
+----- first round -----+      +---- second round --...
+------- specified time -------+
.Ed
.It Fl E
don't delete the SA even if the SA looks dead.
.It Fl S
don't tell syslog(8) what the SA has been deleted.
.It Fl d
debug mode.
.\"
.\".Sh RETURN VALUES
.\"
.\".Sh FILES
.\"
.Sh SEE ALSO
.Xr setkey 8 ,
.Xr syslogd 8
.\"
.\".Sh HISTORY
.\"
.Sh BUGS
.Nm
can only detect the dead SAs which based an IP address.
For example, when the SA is only for a TCP session,
.Nm
cannot detect if the SA is dead or alive because
.Nm
uses ICMP packets.
the peer sends back a response even if the SA is dead.


CHANGELOG
$KAME: CHANGELOG,v 1.2458 2003/08/18 05:39:09 itojun Exp $

<200308>
Sun Aug 17 JST 2003  itojun@iijlab.net
	* sys/netinet{,6}/in{,6}_proto.c: enforce inbound IPsec policy on
	  protocols that falls into raw socket as a wildcard case.

2003-08-15  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/pim6sd/{cfparse.y, mrt.[ch], pim6_proto.c, rp.c}
	Now static-RP configuration can work together with bootstrap-
	message-based RP configuration

2003-08-10  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/{freebsd5, kame}
	sync with FreeBSD-5.1 RELEASE (behaviour is not confirmed yet)

2003-08-07  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* kame/kame/had/mpa.c,haadisc.c
	A home address should be passed in the source address of a mobile
	solicit pakcet. This problem was reported and donated the code by
	Kenichi Yajima <yajima@netlab.nec.co.jp>
	
2003-08-07  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/cftoken.l: use a stronger lexical check for
	interface names.
	From: Hajimu UMEMOTO <ume@mahoroba.org>
	* kame/kame/dhcp6/config.c (configure_interface): check interfaces
	specified in the configuration file really exist.

2003-08-05  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* removed vrrp6 support because it might be a patent infringement.
	  http://www.ietf.org/ietf/IPR/cisco-ipr-draft-ietf-vrrp-ipv6-spec.txt

2003-08-01  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: implemented various clarifications and some new
	  features:
	- supported DNS name and NTP servers options
	- dhcp6s now uses the server/relay port to send a relay-rely
	  message
	- dhcp6s now ignores a rebind message when it cannot find a
	  binding, according to prefix-delegation-04
	- dhcp6relay now accepts relay-reply messages on the socket
	  listening on the server/relay port
	* The DHCPv6 implementation is now built by default under
	  *bsd/usr.sbin/.

<200307>
Thu Jul 31 19:19:48 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c,mip6_cncore.c
	sending/receiving a binding refresh request message is supported.

Tue Jul 29 17:09:15 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.[hc]
	select a CoA using the source address selection like algorithm.
	suggested by Francis.Dupont@enst-bretagne.fr.

2003-07-28  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* sys/netinet6/{ip6_output.c, ip6_var.h, mip6_cncore.c,
	mip6_hacore.[ch], nd6.c}:
	CN or HA dosen't add a type2 routing header to a Binding
	Acknowledgement packet when a deregistration packet was issued
	on home link and it was failed by some reasons.
	reported by Yukiyo.Akisada@jp.yokogawa.com and v6pc cert.
	team.

Mon Jul 28 20:03:46 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_cncore.c,mip6_mncore.[hc],in6.c,nd6_rtr.c
	  kame/sys/net/if_hif.[hc]
	assign a CoA separately for each home interface, since a home
	interface is logically independent from each other.

Mon Jul 28 14:38:11 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_hacore.c,mip6_cncore.c:
	- fixed a bug not sending a binding ack when DAD for a link-local
	  address is failed.  (en-bugged during re-structuring BC logic.)
	- fixed a bug using RTHDR2 with Mobility Headers other than
	  a binding ack.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

Fri Jul 25 19:12:15 JST 2003  itojun@iijlab.net
	* sys/netinet/ah_core.c: hmac-ripemd160 support

Fri Jul 25 18:07:43 JST 2003 sakane@kame.net
	* kame/sys/netkey/key.c
	fixed that the kernel crashed when key_spdacquire() was called
	because key_spdacquire() had been implemented imcopletely.

Thu Jul 24 16:11:06 JST 2003	keiichi@iij.ad.jp
	* mip6 related files.
	simplify the prefix and advertising router list management
	mechanism on a mobile node.

2003-07-20  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/config.c (configure_duid): fixed a bug of
	allocating short memory.  Upgrading is required if you configure
	dhcp6s with the duid statement.

Sat Jul 19 18:12:01 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_aesctr.c: support draft-ietf-ipsec-ciph-aes-ctr-03
	  as there's no official DOI assignment yet, we use private DOI number.
	* sys/netinet6/ah_aesxcbcmac.c: support
	  draft-ietf-ipsec-ciph-aes-xcbc-mac-03.
	  as there's no official DOI assignment yet, we use private DOI number.

Tue Jul 15 20:02:05 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_rijndael.c: simplify and update rijndael code.
	  markus@openbsd

2003-07-14  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: revised the relay agent implementation
	based on dhcpv6-28.
	- dhcp6relay now relays packets between clients and servers using
	  relay forward/reply messages.
	- dhcp6s now accepts relay forward messages and replies with relay
	  reply messages.

Sat Jul 12 15:05:23 JST 2003	suz@crl.hitachi.co.jp
	* openbsd/sys/netinet/igmp_var.h
	makes IGMPV3-kernel compilable on openbsd.

Thu Jul 10 21:35:04 JST 2003	keiichi@iij.ad.jp
	* kame/kame/had/halist.c
	fixed a bug that a DHAAD reply message only includes one global
	address per home agent.  the message must include all global
	addresses of all home agents.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

2003/07/10 21:07:50 JST kjc@csl.sony.co.jp
	the first step for transition to pf/altq.
	marge pf/altq into the existing altq-3.
	allow both pf/altq and altq-3 to coexit.
	currently, ALTQ3_COMPAT and ALTQ3_CLFIER_COMPAT are defined
	in sys/altq/altq.h to support altq-3.
	altq-3 will remain in kame for research experiments but will
	not be merged into bsd releases.

2003-07-08  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* kame/sys/netinet6/{in6_ifattach.c,ip6_mroute.c,ip6_mroute.h}
	protect against interface removal.

Tue Jul  8 18:51:06 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c
	insert alternate careof address sub-option when creating a
	binding update for the home registration.  This sub-option is
	MUST when we protect the message with ESP and strictly speaking,

kame/kame/dsadd/dsadd.pl
#! @LOCALPREFIX@/bin/perl
# $Id: dsadd.pl,v 1.1 2003/01/10 03:46:47 sakane Exp $
# Dead SA Detection

require 'getopts.pl';

$ping6 = '/sbin/ping6';
$ping = '/sbin/ping';
$setkey = '/usr/sbin/setkey';
$logger = '/usr/bin/logger';

$opt_c = 2;	# see -c option.
$opt_E = 0;	# see -E option.
$opt_S = 0;	# see -S option.
$opt_i = 300;	# see -i option.
$opt_L = 0;	# see -L option.

sub usage
{
	print "Usage: dsadd [-c number] [-i time] [-LESvd]\n";
	print "\t-c: specify the number of icmp echo packet to be sent.\n";
	print "\t-L: force to check only once.\n";
	print "\t-i: if the current round takes more than the specified\n";
	print "\t    time in seconds the next round will start immediately\n";
	print "\t    after the current will finish.  otherwise the next will\n";
	print "\t    start after the specified time from the current first\n";
	print "\t    check.\n";
	print "\t-E: don't delete the SA even if the SA looks dead.\n";
	print "\t-S: don't tell syslog(8) what the SA has been deleted.\n";
	print "\t-d: debug mode.\n";
	exit 0;
}

do Getopts('c:i:LESdh');
&usage if ($opt_h);
$debug++ if ($opt_d);

select(STDOUT); $| = 1;
open(OLDERR, ">&STDERR");
open(STDERR, ">&STDOUT");

AGAIN:
%sa_dst = ();
$nextround = time + $opt_i;

%sa_dst = &getsa;

foreach $k (keys %sa_dst) {

	print "\nchecking $k\n" if ($debug);

	if (&pinging($k, 0)) {
		print "  ### seems dead.\n" if ($debug);
		&delete_sa($k) if (! $opt_E);
		next;
	}

	print "  ### seems alive.\n" if ($debug);
}

$interval = $nextround - time;

exit 0 if ($opt_L);

if ($interval > 0) {
	print "\nthe next round will start ${interval}(s) after.\n" if ($debug);
	sleep $interval;
}

goto AGAIN;

#
sub getsa
{
	my ($key, $src, $dst, $proto, $mode, $spi);
	my %sa;

	open(IN, "$setkey -D |");
	while (<IN>) {
		if (/^[^\t]/) {
			($src, $dst) = split(/\s+/, $_);
		} elsif (/^\t(esp|ah) mode=(\w+) spi=\d+\((0x[0-9a-f]+)\)/) {
			$proto = $1;
			$mode = $2;
			$spi = $3;
			$key = "$src $dst $proto $spi $mode";
			$sa{$key} = "$dst";
		}
	}
	close(IN);

	return %sa;
}

sub pinging
{
	my ($key, $need_policy) = @_;
	my ($pingcmd, $policy, $dead, $check);

	$check = 0;
	$dead = 0;

	if ($need_policy) {
		$policy = "-P 'out none'";
	} else {
		$policy = '';
	}

	if ($sa_dst{$key} =~ /:/) {
		$pingcmd = $ping6;
	} else {
		$pingcmd = $ping;
	}

	print "  $pingcmd -nq -c $opt_c $policy $sa_dst{$key}\n" if ($debug);
	open(IN, "$pingcmd -nq -c $opt_c $policy $sa_dst{$key} |");
	while (<IN>) {
		print "  $_" if ($debug && !/^$/);
		if (/100% packet loss/) {
			$dead++;
			last;
		}
		$check++;
	}
	close(IN);

	$dead++ if (!$check);

	return $dead;
}

sub delete_sa
{
	my ($key) = @_;
	my ($cmd, $src, $dst, $proto, $spi, $mode);

	eval "(\$src, \$dst, \$proto, \$spi, \$mode) = qw/$key/";
	$cmd = sprintf "delete $src $dst $proto $spi;";
	print "  $cmd\n" if ($debug);

	system("$logger -t dsadd \"SA deleted src=$src dst=$dst proto=$proto spi=$spi mode=$mode\"") if (! $opt_S);

	open(OUT, "| $setkey -c");
	print OUT "$cmd\n";
	close(OUT);

	return;
}
